src/theory/sets/infer_proof_cons.cpp

```diff
@@ -0,0 +1,333 @@
+      // (and (set.member x S) (= S (op T1 T2))) =>
+      // rewrite((set.member x (op T1 T2)))
+      // this holds by applying the equality as a substitution to the first
+      // assumption and rewriting.
+      std::vector<Node> exp(assumps.begin() + 1, assumps.end());
+      Node aelim = psb.applyPredElim(assumps[0], exp);
+      success = (aelim == conc);
+      // should never fail
+      Assert(success);
+    }
+    break;
+    case InferenceId::SETS_UP_CLOSURE:
+    case InferenceId::SETS_UP_CLOSURE_2:
+    {
+      NodeManager* nm = nodeManager();
+      // An example inference is:
+      // (set.member x A) ^ (set.member y B) ^ (= x y) => (set.member x k)
+      // where k is the purification skolem for (set.inter A B).
+      Assert(conc.getKind() == Kind::SET_MEMBER);
+      Node so = SkolemManager::getUnpurifiedForm(conc[1]);
+      Trace("sets-ipc") << "Unpurified form " << so << std::endl;
+      // We first compute the single step rewriting of the conclusion.
+      // For the above example, memor would be:
+      // (and (set.member x A) (set.member x B)).
+      Node memo = nm->mkNode(Kind::SET_MEMBER, conc[0], so);
+      Node memor = d_tsr->rewriteMembershipBinaryOp(memo);
+      Trace("sets-ipc") << "Single step rewriting of membership " << memor
+                        << std::endl;
+      Assert(memo != memor);
+      // collect the memberships in the premise
+      std::vector<Node> assumpMem;
+      std::vector<Node> assumpOther;
+      // We now partition the antecedant to the membership
+      // part (assumpMem) and the substitution part (assumpOther). The
+      // membership part will be equivalent via rewriting to the conclusion.
+      for (const Node& a : assumps)
+      {
+        Node aa = a.getKind() == Kind::NOT ? a[0] : a;
+        if (aa.getKind() == Kind::SET_MEMBER)
+        {
+          assumpMem.push_back(a);
+        }
+        else
+        {
+          assumpOther.push_back(a);
+        }
+      }
+      Assert(assumpMem.size() == 1 || assumpMem.size() == 2);
+      Node msrc;
+      // Use AND_INTRO to put the memberships together if necessary.
+      if (assumpMem.size() == 2)
+      {
+        msrc = nm->mkAnd(assumpMem);
+        psb.addStep(ProofRule::AND_INTRO, {assumpMem}, {}, msrc);
+      }
+      else
+      {
+        msrc = assumpMem[0];
+      }
+      // Now, prove the equivalence between the memberships and the
+      // conclusion, possibly using the substituion in assumpOther.
+      bool isOr = (memor.getKind() == Kind::OR);
+      size_t ntgts = isOr ? 2 : 1;
+      for (size_t i = 0; i < ntgts; i++)
+      {
+        Node mtgt = isOr ? memor[i] : memor;
+        Trace("sets-ipc") << "...try target " << mtgt << std::endl;
+        if (psb.applyPredTransform(msrc, mtgt, assumpOther))
+        {
+          success = true;
+          if (isOr)
+          {
+            // if union or minus, we get the desired (left or right) conclusion
```

Comment:
Why minus is included here? The function `rewriteMembershipBinaryOp`(rewriteMembershipBinaryOp) only introduces `OR` for `union` [here](https://github.com/cvc5/cvc5/blob/main/src/theory/sets/theory_sets_rewriter.cpp#L660).

Reply:
MINUS is included here, I will add an assertion to make this clearer.

Reply:
MINUS is included in the rewrite function, but there is no OR term in the conclusion. 
minus
 `(set.member x (set.minus A B))` is rewritten as `(and (set.member x A) (not (set.member x B)))`
union
 `(set.member x (set.union A B))` is rewritten as `(or (set.member x A) (set.member x B))`